src/AutofacSerilogIntegration/ContextualLoggingModule.cs

```diff
@@ -46,7 +50,7 @@ protected override void Load(ContainerBuilder builder)
             {
                 builder.Register(c =>
                 {
-                    LoggerProvider provider = new LoggerProvider(_logger);
+                    LoggerProvider provider = new LoggerProvider(_loggerLambda(c));
```

Comment:
You have to understand that this weill make the logger effectively singleton, so no configuration reloads to which you allude in your example.

Reply:
Indeed. In this scenario I would get rid of LoggerProvider altogether as it solves the same goal as the lamda does. But this would be a too bold change to propose:)

Reply:
Even if you get rid of the logger provider (that, by itself, is not a big issue), what will you do with the lifetime?

Reply:
This is a good point. In the example above `SerilogConfiguratorFromSettings` should implement IDisposable and take care or logger field lifetime. Probably alternate approach would be: use LoggerConfiguration instance or lambda as it seems to be a lightweight object so GC can take care of it if needed.

Reply:
> In the example above `SerilogConfiguratorFromSettings` should implement IDisposable and take care or logger field lifetime.

This will take care of the disposal itself, but (a) `SerilogConfiguratorFromSettings` cannot have any `ILogger`-consuming dependencies and (b) you still get only single-instance behaviour. Or, more specifically, you _can_ try implementing per-scope, but the state of your loggers will be very unpredictable.

Reply:
> SerilogConfiguratorFromSettings cannot have any ILogger-consuming dependencies 

Indeed, here it is not possible. In scenario where you have more control over DI (wcf project for instance) - it can done.,
log4net sample:
```

using System;
using Autofac;
using JetBrains.Annotations;
using log4net;


/// <summary>
/// Called from either global.asax or AutofacServiceHostFactory
///     //configsProvider is similar to IConfigurationSource in .net core
///     var configsProvider = GetRegularConfigsProvider(); 
///     var bootstrap = new Bootstrap();
///     bootstrap.Run(configsProvider);
/// </summary>
public class Bootstrap {

    #region properties injected by DI

    /// <summary> Logger; can be null </summary>
    [CanBeNull] private ILog _log;

    /// <summary> Log is set by IoC Container </summary>
    [CanBeNull]
    public ILog Log {
        get => _log;
        set => _log = value;
    }

    /// <summary> ссылка на родительский контейнер autofac </summary>
    public ILifetimeScope Container { get; set; }

    #region конфирураторы настроек и соотв настройки - injected by autofac (one-by-one)

    public Log4netConfigurator ConfigLog4net { get; set; }

    public ServiceSettingsConfigurator ConfigServiceSettings { get; set; }

    #endregion конфирураторы настроек и соотв настройки - injected by autofac (one-by-one)

    public AutofacConfiguration ConfigAutofac { get; set; }

    #endregion properties injected by DI

    public void Run(ConfigsProvider configsProvider) {
        try {
            #region build initial autofac container

            //build initial autofac container and start chicken(egg) nested resolution
            if (Container == null) {
                _log?.Info("building initial autofac container...");
                var builder = new ContainerBuilder();
                builder.RegisterType<Bootstrap>().AsSelf().PropertiesAutowired();
                builder.RegisterInstance(configsProvider).As<ConfigsProvider>();
                Container = builder.Build();
                var nextBootstrap = Container.Resolve<Bootstrap>();
                if (nextBootstrap == null) throw new NullReferenceException(nameof(nextBootstrap));
                nextBootstrap.Run(configsProvider);
                return;
            }

            #endregion


            #region log4net

            //log4net
            if (ConfigLog4net == null) {
                _log?.Info("log4net init...");
                var scope = Container.BeginLifetimeScope(s => {
                    s.RegisterType<Log4netConfigurator>().AsSelf().SingleInstance();
                    s.RegisterModule<Log4netModule>();
                });
                ConfigLog4net = scope.Resolve<Log4netConfigurator>(); //force XmlConfigurator.Configure(); on the instance creation
                if (ConfigLog4net == null) throw new NullReferenceException(nameof(ConfigLog4net));
                var nextBootstrap = scope.Resolve<Bootstrap>();
                if (nextBootstrap == null) throw new NullReferenceException(nameof(nextBootstrap));
                nextBootstrap.Run(configsProvider);
                return;
            }

            if (_log == null) throw new NullReferenceException(nameof(_log));

            #endregion log4net


            #region ServiceSettings

            //ServiceSettings
            if (ConfigServiceSettings == null) {
                _log?.Info("creating ServiceSettingsConfigurator...");
                var scope = Container.BeginLifetimeScope(s => {
                    s.RegisterType<ServiceSettingsConfigurator>().AsSelf().SingleInstance();
                    s.RegisterModule<Log4netModule>(); //Log4netModule have to be registered on each inner scope
                });
                ConfigServiceSettings = scope.Resolve<ServiceSettingsConfigurator>();
                if (ConfigServiceSettings == null) throw new NullReferenceException(nameof(ConfigServiceSettings));
                var nextBootstrap = scope.Resolve<Bootstrap>();
                if (nextBootstrap == null) throw new NullReferenceException(nameof(nextBootstrap));
                nextBootstrap.Run(configsProvider);
                return;
            }

            #endregion ServiceSettings


            #region config autofac globally

            //config autofac globally
            if (ConfigAutofac == null) {
                _log?.Info("configuring autofac...");
                var scope = Container.BeginLifetimeScope(s => {
                    s.RegisterType<AutofacConfiguration>().AsSelf().PropertiesAutowired();
                    s.RegisterModule<Log4netModule>(); //Log4netModule have to be registered on each inner scope
                });
                ConfigAutofac = scope.Resolve<AutofacConfiguration>();
                if (ConfigAutofac == null) throw new NullReferenceException(nameof(ConfigAutofac));
                ConfigAutofac.Configure();
            }

            #endregion config autofac globally
        }
        catch (Exception ex) {
            _log?.Error("init error", ex);
            throw;
        }
    }
}
```

Reply:
Unfortunately, this example is too wordy for me to understand the particular points you're trying to make of it.

In `Microsoft.Extensions.Hosting`-based projects (and .net Core Worker service seem to be one of those) there's quite a lot of container control. From [the looks of it](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/host/hosted-services?view=aspnetcore-3.1&tabs=visual-studio&WT.mc_id=-blog-shchowd#worker-service-template) you can do whatever you want as long as you return an `IHostBuilder` that can build an `IHost`.

Reply:
The point is that you can build a container incrementally, onion-way. The key is to use .PropertiesAutowired() and recursive calls.

Reply:
There's nothing that stops you from doing the same in your own `IHostBuilder` or `IHostedService`.